CRM/Core/BAO/UFField.php

```diff
@@ -42,6 +42,59 @@ class CRM_Core_BAO_UFField extends CRM_Core_DAO_UFField {
   private static $_contriBatchEntryFields = NULL;
   private static $_memberBatchEntryFields = NULL;
 
+  /**
+   * Create UFField object.
+   *
+   * @param array $params
+   *   Array per getfields metadata.
+   *
+   * @throws CRM_Core_Exception
+   *
+   * @return CRM_Core_BAO_UFField
+   */
+  public static function create(&$params) {
+    // CRM-14756: kind of a hack-ish fix. If the user gives the id, uf_group_id is retrieved and then set.
+    if (isset($params['id']) && !isset($params['uf_group_id'])) {
+      $groupId = civicrm_api3('UFField', 'getvalue', array(
+        'return' => 'uf_group_id',
+        'id' => $params['id'],
+      ));
+    }
+    else {
+      $groupId = CRM_Utils_Array::value('uf_group_id', $params);
+    }
+
+    if (isset($params['field_name'])) {
+      $field_name = CRM_Utils_Array::value('field_name', $params);
+
+      if (strpos($field_name, 'formatting') !== 0 && !CRM_Core_BAO_UFField::isValidFieldName($field_name)) {
+        throw new CRM_Core_Exception('The field_name is not valid');
+      }
+    }
+
+    if (!(CRM_Utils_Array::value('group_id', $params))) {
+      $params['group_id'] = $groupId;
+    }
+
+    $fieldId = CRM_Utils_Array::value('id', $params);
+
+    if (CRM_Core_BAO_UFField::duplicateField($params)) {
+      throw new CRM_Core_Exception("The field was not added. It already exists in this profile.");
+    }
+    // @todo why is this even optional? Surely weight should just be 'managed' ??
+    if (CRM_Utils_Array::value('option.autoweight', $params, TRUE)) {
+      $params['weight'] = CRM_Core_BAO_UFField::autoWeight($params);
```

Comment:
Probable cause of failure. Field name validation is working correctly now (hopefully).